homeassistant/components/netatmo/sensor.py

```diff
@@ -112,84 +95,136 @@
+            return []
+
+        data_class = data[data_class_name]
+
+        for station_id in data_class.stations:
+            for module_id in data_class.get_modules(station_id):
+                all_module_infos[module_id] = data_class.get_module(module_id)
+
+            all_module_infos[station_id] = data_class.get_station(station_id)
+
         entities = []
         for module in all_module_infos.values():
-            _LOGGER.debug("Adding module %s %s", module["module_name"], module["id"])
-            for condition in data.station_data.monitoredConditions(
-                moduleId=module["id"]
+            if "_id" not in module:
+                _LOGGER.debug("Skipping module %s", module.get("module_name"))
+                continue
+
+            _LOGGER.debug(
+                "Adding module %s %s", module.get("module_name"), module.get("_id"),
+            )
+            for condition in data_class.get_monitored_conditions(
+                module_id=module["_id"]
             ):
-                entities.append(NetatmoSensor(data, module, condition.lower()))
+                entities.append(
+                    NetatmoSensor(
+                        data_handler, data_class_name, module, condition.lower()
+                    )
+                )
+
         return entities
 
-    def get_entities():
+    async def get_entities():
         """Retrieve Netatmo entities."""
         entities = []
 
-        for data_class in [pyatmo.WeatherStationData, pyatmo.HomeCoachData]:
-            try:
-                dc_data = data_class(auth)
-                _LOGGER.debug("%s detected!", NETATMO_DEVICE_TYPES[data_class.__name__])
-                data = NetatmoData(auth, dc_data)
-            except pyatmo.NoDevice:
-                _LOGGER.debug(
-                    "No %s entities found", NETATMO_DEVICE_TYPES[data_class.__name__]
-                )
-                continue
-
-            entities.extend(find_entities(data))
+        for data_class_name in [
+            WEATHERSTATION_DATA_CLASS_NAME,
+            HOMECOACH_DATA_CLASS_NAME,
+        ]:
+            entities.extend(await find_entities(data_class_name))
 
         return entities
 
-    async_add_entities(await hass.async_add_executor_job(get_entities), True)
+    async_add_entities(await get_entities(), True)
 
     @callback
-    def add_public_entities():
+    async def add_public_entities(update=True):
         """Retrieve Netatmo public weather entities."""
-        entities = []
-        for area in entry.options.get(CONF_WEATHER_AREAS, {}).values():
-            data = NetatmoPublicData(
-                auth,
-                lat_ne=area[CONF_LAT_NE],
-                lon_ne=area[CONF_LON_NE],
-                lat_sw=area[CONF_LAT_SW],
-                lon_sw=area[CONF_LON_SW],
+        temp_data_classes = []
+        entities = {
+            device.name: device
+            for device in async_entries_for_config_entry(
```

Comment:
It looks like we hold an entire copy of the device registry in memory here at the time this is setup.  Can we get these another way ?

Reply:
Well, yeah, actually I only need the name and idea. Fixed that.